internal/kafka/internal/services/data_plane_kafka.go

```diff
@@ -150,30 +151,57 @@ func (d *dataPlaneKafkaService) processRealKafkaDeployment(ks *dbapi.DataPlaneKa
 		log.Warningf("clusterId for kafka cluster %s does not match clusterId. kafka clusterId = %s :: clusterId = %s", kafka.ID, kafka.ClusterID, cluster.ClusterID)
 		return
 	}
+
+	// Notes on state transitions
+	//  - 'suspending' state is set by the user. This must only transition to 'resuming', 'suspended' or 'deprovision'.
+	//     - FSO will only send the status 'suspended' if the Kafka instance was already in a 'suspending' or 'suspended' state.
+	//     - FSO will not change the status of the ManagedKafka CR prior to transitioning to 'suspended' unless an error occurs.
+	//       e.g. If the Kafka instance was in a 'ready' state previously, FSO will keep reporting its status as 'ready' until it
+	//            finally reports 'suspended'. If an error occurs, the 'Ready' condition of the CR will have the values 'Status=False,Reason=Error'.
+	//  - 'resuming' state is set by the user from a 'suspending' or 'suspended' state. This must never transition to 'suspended'.
+	//  - 'suspended' state must only transition to 'resuming' or 'deprovision'.
+	//  - 'deprovision' state is set by the user. This must only transition to 'deleting' or 'failed'.
+	//     - FSO will only send the status 'deleted' if kafka was already in a 'deprovisioning' state.
+	//  - 'failed' (or 'error') state may occur at any time. However, KFM must never transition a Kafka instance to 'failed' from 'suspending' and 'suspended' states.
 	var e *serviceError.ServiceError
 	switch s := d.getManagedKafkaStatus(ks); s {
 	case statusReady:
-		// Store the routes (and create them) when Kafka is ready. By the time it is ready, the routes should definitely be there.
-		e = d.persistKafkaRoutes(kafka, ks, cluster)
-		if e == nil {
-			kafka.AdminApiServerURL = ks.AdminServerURI
-			e = d.setKafkaClusterReady(kafka)
+		if kafka.Status != constants2.KafkaRequestStatusSuspending.String() {
+			// Store the routes (and create them) when Kafka is ready. By the time it is ready, the routes should definitely be there.
+			e = d.persistKafkaRoutes(kafka, ks, cluster)
+			if e == nil {
+				kafka.AdminApiServerURL = ks.AdminServerURI
+				e = d.setKafkaClusterReady(kafka)
+			}
 		}
 	case statusInstalling:
-		// Store the routes (and create them) if they are available at this stage to lessen the length of time taken to provision the Kafka.
-		// The routes list will either be empty or complete.
-		e = d.persistKafkaRoutes(kafka, ks, cluster)
+		if kafka.Status != constants2.KafkaRequestStatusSuspending.String() {
+			// Store the routes (and create them) if they are available at this stage to lessen the length of time taken to provision the Kafka.
+			// The routes list will either be empty or complete.
+			e = d.persistKafkaRoutes(kafka, ks, cluster)
```

Comment:
@MikeEdgar can the routes change once a Kafka is resumed? Right now, we don't support overwriting the routes that's already persisted in the database. Do we need to accommodate for this?

Reply:
The routes will remain unchanged and will continue to be published in the status.

Reply:
Thanks for confirming @MikeEdgar!

Reply:
@JameelB - after some additional testing, I've found that the routes are being set to an empty array in the status while the MK is suspended. The reason is that kas-fleetshard operator uses the Kafka broker pods to determine the AZ for each route, and they are not available when suspended, of course. After resuming the cluster, the routes will again be populated. Please let me know if this could be a problem on the control plane side.

Reply:
Thanks for the update @MikeEdgar. I'll have a look to check if this causes a problem on our side.